pkg/controller/bundle/bundle_unpacker.go

```diff
@@ -75,7 +76,12 @@ func (c *ConfigMapUnpacker) job(cmRef *corev1.ObjectReference, bundlePath string
 					Name: cmRef.Name,
 				},
 				Spec: corev1.PodSpec{
-					RestartPolicy:    corev1.RestartPolicyOnFailure,
+					// With restartPolicy = "OnFailure" when the spec.backoffLimit is reached, the job controller will delete all
+					// the job's pods to stop them from crashlooping forever.
+					// By setting restartPolicy = "Never" the pods don't get cleaned up since they're not running after a failure.
+					// Keeping the pods around after failures helps in inspecting the logs of a failed bundle unpack job.
+					// See: https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-backoff-failure-policy
+					RestartPolicy:    corev1.RestartPolicyNever,
```

Comment:
This is all well and good for when a pod's container errors out after it's running (e.g invalid bundle manifests). In that case the job fails fast due to `BackOffLimit=4`, the InstallPlan is marked as failed, and the failed pod sticks around for inspection.

However if the pod never starts running i.e `phase=Pending` and encounters the `ImagePullBackoff` or `ErrImagePull` error then the job won't fail fast. 
The job's `backoffLimit` is never invoked since the container hasn't exited in error yet.
So the job keeps running and will only fail once we hit the larger (10m) timeout from `ActiveDeadlineSeconds` at which point the pod is terminated.
Even more unfortunately the pod is removed after the timeout so you can't inspect it afterwards.

So for the case of a non-existent image in a bundlelookup, we're not really failing fast, or preserving the pod for debugging afterwards. We'll just eventually have the InstallPlan as Failed after a long timeout.

Reply:
@benluddy I'm not sure how we want to handle this case. We could try and inspect the pod state for the case when we have an image pull failure and then fail fast ourselves:
```YAML
apiVersion: v1
kind: Pod
metadata:
  name: 8c3e265337115bdeff1ab7a092b9f531bffc580830a021ffe04d5214ee4k7kf
  namespace: default
  ownerReferences:
  - apiVersion: batch/v1
    blockOwnerDeletion: true
    controller: true
    kind: Job
    name: 8c3e265337115bdeff1ab7a092b9f531bffc580830a021ffe04d5214eea1419
    uid: 273dfba6-426e-46e1-b586-fdae4b792ad4
status:
  initContainerStatuses:
  ...
  - image: quay.io/foo/bar:latest
    imageID: ""
    lastState: {}
    name: pull
    ready: false
    restartCount: 0
    state:
      waiting:
        message: 'rpc error: code = Unknown desc = failed to pull and unpack image
          "quay.io/foo/bar:latest": failed to resolve reference "quay.io/foo/bar:latest":
          unexpected status code [manifests latest]: 401 UNAUTHORIZED'
        reason: ErrImagePull
  phase: Pending
```

Seems a little brittle though.

Reply:
Well, a long timeout is still an improvement over no timeout.

Is there a middle ground where the Pending lookup condition message indicates what is currently happpening? Then it would take a while to fail, but at least if a user inspects the status then she will see that there is a Job with a pod in Pending/ImagePullBackOff. I agree that there may be more we can do to fail fast in special cases, but also agree that it could be brittle. That could be a future improvement?

Reply:
We can peek at the container statuses while we're waiting for bundle lookup to try and add more context to the `BundleLookupPending` condition.
Not sure if that covers anything else but at least it will surface this particular issue to the user. Going to add that in the next update.

Reply:
Okay so the BundleLookupPending condition should have the container status info for any containers that aren't ready in pending unpack Job pods.
That way while the InstallPlan is stalled on phase Installing with BundleLookupPending for a non-existent bundle image we can see the reason why:
```YAML
apiVersion: operators.coreos.com/v1alpha1
kind: InstallPlan
status:
  bundleLookups:
    conditions:
    - lastTransitionTime: "2021-04-09T00:44:55Z"
      message: "unpack job not completed: Unpack pod(default/8c3e265337115bdeff1ab7a092b9f531bffc580830a021ffe04d5214eenb2xr)
        container(pull) is pending. Reason: ErrImagePull, Message: rpc error: code
        = Unknown desc = failed to pull and unpack image \"quay.io/foo/bar:latest\":
        failed to resolve reference \"quay.io/foo/bar:latest\": unexpected status
        code [manifests latest]: 401 UNAUTHORIZED \n"
      reason: JobIncomplete
      status: "True"
      type: BundleLookupPending
    path: quay.io/foo/bar:latest
``` 

Of course it keeps toggles back and forth between `ErrImagePull` and `ImagePullBackOff` as it backs off but that's fine I think:
```YAML
apiVersion: operators.coreos.com/v1alpha1
kind: InstallPlan
status:
  bundleLookups:
    conditions:
    - lastTransitionTime: "2021-04-09T00:44:55Z"
      message: "unpack job not completed: Unpack pod(default/8c3e265337115bdeff1ab7a092b9f531bffc580830a021ffe04d5214eenb2xr)
        container(pull) is pending. Reason: ImagePullBackOff, Message: Back-off pulling
        image \"quay.io/foo/bar:latest\" \n"
      reason: JobIncomplete
      status: "True"
      type: BundleLookupPending
    path: quay.io/foo/bar:latest
  phase: Installing
```

Hopefully that makes the non-existent image case more apparent.